### PR TITLE
Do not observe visible hydrate components more than once

### PR DESCRIPTION
**Pull request by FredKSchott**
## Changes

There was an issue on the docs site where two islands of the same component, hydrated with `client:visible`, caused the `onVisible` hook to observe both components multiple times. This caused some sort of conflict between the two and incorrect results for `entry.isIntersecting`. The result was that the visible theme toggle wasn't hydrating on page load with `client:visible`.

## Testing

- Covered by existing `client:visible` test.
- Can't currently test this specific case in our test suite since it requires browser behavior.

## Docs

- N/A

### CODE REVIEW

**Comment by jasikpark** [comment]
Wouldn't a different solution be to `.unobserve()` the element on hydrate? Since it doesn't need to be tracked by the hydration IntersectionObserver anymore?

**Comment by matthewp** [comment]
Yeah, I think we should use `.unobserve()` if that will work.

**Comment by FredKSchott** [comment]
The issue I was seeing was that `entry.isIntersecting` was always false, so the "hydrate this code" flow after that check was never actually hit. I assume that's where any `unobserve` code would go

**Comment by matthewp** [comment]
@FredKSchott Oh, so is the problem here that the astro-id is not unique? Those are in fact not the same "instance" of the component, correct? They should render independently.

**Comment by jasikpark** [comment]
Yep, they're two different 'client:visible' elements.

**Comment by FredKSchott** [comment]
Correct, there is an existing issue where astro-root `uid` values are hashed to the file itself, and not the individual components on the page. Is the idea to fix that instead? I kind of wondered if that might be intentional

**Comment by jasikpark** [comment]
I feel like it only makes sense for those to be loaded at the same time if they are pure functions or something? I think the default should definitely be to load them individually though

**Comment by FredKSchott** [comment]
Fixed! `uid` value is now completely unique per island. @matthewp can you confirm that this is good? Hard to see intention in the original code, and why a hash of the stringified props was used.

**Comment by jasikpark** [comment]
The hash of stringified props was added in a PR to do something similar to this PR, there was a bug where components even with different props were SSR with the first component's props.

**Comment by FredKSchott** [comment]
Ah, that makes sense. This is ready for review then! /cc @jasikpark @matthewp

**Comment by matthewp** [comment]
Would like @natemoo-re  to weigh in as he did the implementation of the astro id. I thought it was to make the ids deterministic (maybe so tests would work?). I'm not sure the reason exactly.

**Comment by natemoo-re** [comment]
Thanks for the ping @matthewp! I am -1 on using `nanoid` here, we should stick with a hashing method.

The reason for using `shorthash` and basing the `uid` on the contents of the island is that a unique (random) id would effectively cache-bust every page every time `astro build` is run. HTML pages with islands (but no actual content diff) would be invalidated because the island would _always_ have a new of the new, random `uid`. At scale, this could be a major problem.

Is there some other information that we could add to the hash that would sidestep the original issue but maintain the determinism of the current approach? Maybe the current page's path relative to the Astro root?

**Comment by matthewp** [comment]
Ah, I think we need a distinct id for each h() all in the entire __renderPage() call.  This object gets passed down to every component in the tree: https://github.com/snowpackjs/astro/blob/main/packages/astro/src/compiler/index.ts#L184-L192

We could add a `createId()` method to that context object that increments a counter and astro component could then call that. That would be deterministic and would never clash with another component used.

**Comment by FredKSchott** [comment]
+1 for the counter based on my (limited :) understanding of the issue. @natemoo-re wdyt?

**Comment by natemoo-re** [comment]
That idea works for me! I think we'd need to use `Astro.request.toString()` in addition to the counter or they could clash across pages.

**Comment by FredKSchott** [comment]
@matthewp I think @natemoo-re just called you clever 🤭

**Comment by FredKSchott** [comment]
> We could add a createId() method to that context object that increments a counter and astro component could then call that. That would be deterministic and would never clash with another component used.

@matthewp what does this need to be scoped to? A counter for each page? A counter for the entire build / lifetime of the dev server? Just want to confirm before implementing.

**Comment by matthewp** [comment]
@FredKSchott Just to the page right? Hydration is just the page. You can add the function in this object: https://github.com/snowpackjs/astro/blob/main/packages/astro/src/compiler/index.ts#L185-L188

Pseudo-code:

```js
let counter = 0;
// ...
      value: {
        pageCSS: css,
        request,
        createId() {
          return count++;
        }
      },
```

Then to use this in __astro_component you can do:

```js
const id = props[Symbol.for('astro.context')].createId();
```

**Comment by jasikpark** [comment]
I wonder how this'll interact with dynamic routing? I imagine it's something that can wait until that's a problem though.

**Comment by FredKSchott** [comment]
Each individual page from a single dynamic collection (do we still call these collections?) is treated as a different page, so hopefully fine?

Re-reading @natemoo-re 's comment above, this isn't just about determinism but is actually about cache busting? Can you say more about that, is the hash used for fingerprinting in production in some way? 

I'm trying to figure out if it's okay to just hash the `counter` alone, or the `html + counter`. I think we should do `html + counter` if only so that each uid is unique, and we aren't reusing the same `hash(0)`, `hash(1)`, etc. across every page of your site.

**Comment by jasikpark** [comment]
I don't actually have a real consideration for why dynamic routing would change things, it just feels like a place a bug would hide!

**Comment by matthewp** [comment]
@FredKSchott it's ok to just hash the counter alone. I think what @natemoo-re means is that if the content of the HTML changes the `etag` will change. It's not important that the id be unique across the site, because it's only used for hydrating components on a single page. When you travel to another page it will hydrate just that page.

A dynamic route page is still a single page, so it again doesn't matter if the ids are the same.

**Comment by FredKSchott** [comment]
@matthewp PTAL! Fixed with your suggestion above, worked like a charm.

I'd still like uids to be truely unique, otherwise component A on page 1 & component B on page 2 would have the same `uid={hash(0)}` value, even though they are not related at all. Ditto for page 3, 4, 5, etc.  I imagine this will make things will be harder to debug and generally confusing why unique values are so often repeated.

Also if `html` changes but the counter doesn't change, the the etag should still trigger a change. Although, I would assume that by definition an `html` change = a overall page `html` change, and therefore an etag value would change anyway.

File: packages/astro/test/vue-component.test.js [hunk 1]
```diff
@@ -13,9 +13,10 @@ Vue('Can load Vue', async ({ runtime }) => {
...
 
   const $ = doc(result.contents);
-  assert.equal($('h1').text(), 'Hello world!', 'Can use slots');
-  assert.equal($('button').length, 2, 'Can render components');
-  assert.equal($('pre').text(), '5', 'Can render nested components');
+  const allPreValues = $('pre').toArray().map((el) => $(el).text());
+  assert.equal(allPreValues, ['0', '1', '10', '100'], 'renders all 4 components correctly');
+  assert.equal($('astro-root').length, 3, 'renders 3 astro-roots');
+  assert.equal($('astro-root[uid]').length, 3, 'all astro-roots have uid attributes');
 });
```
Comment (jasikpark) [inline packages/astro/test/vue-component.test.js:19]: would it make sense to add a test for multiple elements loading on the same hook + testing that astro id's are unique here? using a `Set()` 😆

**Review by jasikpark — APPROVED** [review]

File: packages/astro/test/vue-component.test.js (hunk 1 above)
Comment (FredKSchott) [reply packages/astro/test/vue-component.test.js:19]: lol yea feels silly not to test that at all :) test added

**Review by jasikpark — APPROVED** [review]

**Review by matthewp — APPROVED** [review]

**Comment by matthewp** [comment]
@FredKSchott, approved.

The whole id system imo should be replaced by a custom element, that way no ids will be need at all:

```html
<astro-hydrate-visible src="/_astro/src/elements/foo.js">
  <div class="preact root here">...</div>
</astro-hydrate-visible>
```

But we need to lift script tags into the head to make this possible. Something for the future.

File: docs/src/components/RightSidebar/MoreMenu.astro [hunk 2]
```diff
@@ -64,5 +64,5 @@ const {editHref} = Astro.props;
   </li>
 </ul>
 <div style="margin: 2rem 0; text-align: center;">
-  <ThemeToggleButton client:idle />
+  <ThemeToggleButton client:visible />
```
Comment (MarcusOtter) [inline docs/src/components/RightSidebar/MoreMenu.astro:67]: This broke the theme toggler

**Review by MarcusOtter — COMMENTED** [review]
~~Seems like this change broke the theme toggler for some reason! Tested in both chrome and firefox.~~
False alarm (see https://github.com/snowpackjs/astro/issues/1146)

File: docs/src/components/RightSidebar/MoreMenu.astro (hunk 2 above)
Comment (MarcusOtter) [reply docs/src/components/RightSidebar/MoreMenu.astro:67]: Err, false alarm 😄

**Pull request merged**